### ISSUE TITLE
Paste from Word - basic formatting support

### ISSUE THREAD

**Issue by f1ames** [ckeditor/ckeditor5#1184]
Extracted from #1003. 

Paste from Word is a huge feature and we will work on it iteratively. As a first batch of features we would like to support was extracted (see https://github.com/ckeditor/ckeditor5/issues/1003#issuecomment-409196061), I have created this task to cover all related stuff to not clutter the parent/umbrella issue too much.

The first batch of features contains (not the final list and may change):

1. bold, italics, underline, strikethrough
1. headings
1. links
1. lists
1. images

First step is to see if any of this already works correctly and also take a quick look at markup generated by different Word versions and some similar software (GDocs, Libre Office) to get broader view.

**Comment by wwalc** [comment]
Not sure how to count support for headings. Technically they "work" but in reality the fact that heading 1 isn't preserved is wrong IMHO. After all what's wrong with using "heading 1" in MS Word documents?
Perhaps during pasting we should detect if there is "H1" used there and automatically convert all headings one level down if H1 is not available. I guess this problem will be another argument for doing https://github.com/ckeditor/ckeditor5-heading/issues/98 sooner than later. It is probably more common to have H1 is Word documents than in content copied from websites.

**Comment by f1ames** [comment]
> Not sure how to count support for headings. Technically they "work" but in reality the fact that heading 1 isn't preserved is wrong IMHO.

I agree. Technically it works (tags/content transformation is fine), but since CKEditor supports `h2`+, `h1` is transformed into paragraph - so from paste from Word point of view it's broken.

Adding support for `h1` in the editor itself (so https://github.com/ckeditor/ckeditor5-heading/issues/98) would solve this issue automatically. If we don't want this approach, PFW plugin could do what @wwalc suggested:

> Perhaps during pasting we should detect if there is "H1" used there and automatically convert all headings one level down if H1 is not available.

and it should solve this issue (the question is if we should always downgrade the headers or only if there is a `h1`?).

Another issue might be that heading transformations list is configurable and might be changed so then downgrading headers may by tricky or result in some unexpected behaviour. Still I agree that converting `h1` to `p` by default is quite confusing.

**Comment by f1ames** [comment]
## Other text editors

Did some quick check with other text editors to catch a glimpse of state of things:

### Pages

_It has quite nice and clear markup when pasted_.

**Works**: 

* bold, italics
* tables
* links
* lists (🎉)

**Broken**
* quote
* images
* underline/strikethrough (these elements are just formatted via styling and not represented in a markup with `<u>` or `<s>` tags)

### Word online

**Works**
* bold, italics
* tables (to some extend, whole content gets bolded)
* links
* images (pasted as `base64` string)

**Broken**
* quotes
* headers (represented only with styles and not `h*` elements)
* list (has structure like `ul > li > p` so instead of list, paragraphs are created)
* underline/strikethrough (same as headers)

### Google Docs

_Both Word online and Google docs generate very similar markup on copy. Structure is more or less correct with a lot of inline CSS._

**Works**
* bold, italics, underline, strikethrough
* tables (to some extend, whole content gets bolded)
* links
* headings (`h1` is transformed into `p`)
* images (with link to online resource)

**Broken**
* quotes
* list (has structure like `ul > li > p` so instead of list, paragraphs are created)

### Libre Office (Writer)

**Works**
* bold, italics, underline, strikethrough
* tables
* links
* headings (`h1` is transformed into `p`)

**Broken**
* quotes
* images
* list (has structure like `ul > li > p` so instead of list, paragraphs are created)

---

In most cases if correct markup is provided (even with lots of styling), the content is transformed correctly by CKEditor (only if given elements have proper converters). Most common issues:

* In some cases elements are represented only with styling and not proper HTML tags (`h*`, `u`, `s`).
* In any of the above editors I haven't seen a dedicated option to insert quotes. Usually it is adviced to use regular `""`, which in most cases are changed to &ldquo; and &rdquo; upon typing. Also this creates inline quote, and CKEditor support block quotes ATM.
* In most editors, it is possible to insert block elements as list items (e.g. headers). CKEditor does not support such structure so it will need some additional handling (important as Word 2016 also allows that so should be kept in mind when working on lists support). That's probably a cause why lists do not work in Google Docs and Libre Office (see above).

**Comment by f1ames** [comment]
Extracted headers issue to https://github.com/ckeditor/ckeditor5-paste-from-word/issues/2.

**Comment by robclancy** [comment]
Is there a tool or something we can suggest people use intermediately? I don't think we can finish our move to ckeditor5 without word pasting unless I can provide an alternative way to do it.

**Comment by robclancy** [comment]
@f1ames our app is a content provider platform. We provide a wide range but articles are the main thing so your general article would be pasted but with images being manual. We also do things like storyboards for video production in tables that would be copied over.

But it would mainly be headings, formatting (like fonts, color, which we would want filtered out) and maybe lists sometimes (they have a habit of just doing `-` with a number in lists which frustrates me a lot). Then tables for the more niche areas that don't need to be perfect.

We are using froala now and moving to ckeditor5 once we can, I have made an add-on already but this paste and some other things you're working on need to be done first. We already use it for things that don't need the more detailed editor (comments for example).

Hopefully after our next milestone I can start to contribute.

**Comment by robclancy** [comment]
Is there a progress update on this? I might be able to look into it more myself in the coming weeks but don't want to be doing something someone who knows better is already doing.

**Comment by Reinmar** [comment]
The first version of the [Paste from Office](https://github.com/ckeditor/ckeditor5-paste-from-office) feature is nearly ready. It will be released with the upcoming release.

If you'd like to test it you can either wait for https://github.com/ckeditor/ckeditor5/issues/1351 to be resolved (PFO will then be on in all samples in the nightly docs: https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/api/index.html) or you can check it out via https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/development-environment.html

**Comment by Reinmar** [comment]
I'm closing this ticket in favor of more specific issues reported in https://github.com/ckeditor/ckeditor5-paste-from-office

**Comment by robclancy** [comment]
Great, thanks.

**Comment by tista-chatterjee** [comment]
![image](https://user-images.githubusercontent.com/91665085/155324201-c30a67cd-f60d-4244-b1b7-18573b3d9c4a.png)

See image:

Bold H1 test - content that has been typed directly into CKEditor (H1).

Non-bold 'Google Docs - Test Content for UAT' - content that has been pasted from Google Docs, then an H1 tag applied to the words.

Expectation would be that all content displays the same, but that is not happening, please help us by fixing this.

**Comment by Namrata99sah** [comment]
Facing the same issue in our environment, kindly help us fix the issue.

**Comment by Suj5432** [comment]
same issue in our case as well as described by @tista-chatterjee. please help fix the issue.

**Comment by amitj-ap** [comment]
When we insert a .emz image in MS Word 2013, copy it, and paste it into CKEditor 5, the image displays correctly. However, if the .emz file is inserted with text in MS Word 2013, and then copied along with the text and pasted into CKEditor 5, only the text appears; the image does not display.

I'm using "@ckeditor/ckeditor5-paste-from-office": "^41.4.2", and plugin - PasteFromOffice

@f1ames @tista-chatterjee, please help the issue fix.

**Comment by wx925401941** [comment]
如果我使用了PasteFromOffice这个插件，从WPS 这款办公软件 复制到这边编辑器中，就会显示全是空白文档
